pkg/processor/build/util/unarchive.go

```diff
@@ -0,0 +1,169 @@
+		} else if filterFile {
+			return nil
+		}
+
+		// copy the file to a new file with the same name in the target path
+		return d.extractFile(file, targetPath)
+	}
+
+	if extractor, ok := format.(archiver.Extractor); ok {
+		if err := extractor.Extract(ctx, input, nil, extractionHandler); err != nil {
+			return errors.Wrap(err, "Failed to extract archive")
+		}
+	}
+
+	return nil
+}
+
+// extractFile extracts a file from the archive to the target path with the same base name
+func (d *Unarchiver) extractFile(file archiver.File, targetPath string) error {
+	filePath := filepath.Join(targetPath, file.NameInArchive)
+
+	// create the directory if it doesn't exist
+	if err := os.MkdirAll(filepath.Dir(filePath), 0755); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed to create directory %s", filepath.Dir(filePath)))
+	}
+
+	// create the new file
+	outFile, err := os.Create(filePath)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed to create file :%s", filePath))
+	}
+	defer outFile.Close() // nolint: errcheck
+
+	// set the file mode
+	err = outFile.Chmod(file.Mode())
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed to change file mode for %s", filePath))
+	}
+
+	// open the file
+	input, err := file.Open()
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed to open file: %s", filePath))
+	}
+
+	// copy the file contents
+	_, err = io.Copy(outFile, input)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed to write file: %s", filePath))
+	}
+	return nil
+}
+
+// filterFile checks if the file should be extracted or not by checking if the destination file path or resolved
+// linked path is outside the target directory
+func (d *Unarchiver) filterFile(file archiver.File, targetPath string) (bool, error) {
+	for _, pathToCheck := range []string{
+		file.NameInArchive,
+		file.LinkTarget,
+	} {
+		if pathToCheck == "" {
+			continue
+		}
+		fullPath, err := filepath.Abs(path.Join(targetPath, pathToCheck))
+		if err != nil {
+			return true, errors.Wrap(err, fmt.Sprintf("Failed to get absolute path for %s", fullPath))
+		}
+		if !strings.HasPrefix(fullPath, targetPath) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// IsArchive checks if the file is an archive
+func IsArchive(source string) bool {
+
+	// Jars are special case
+	if IsJar(source) {
+		return false
+	}
+
+	file, err := os.Open(source)
+	if err != nil {
+		// if we can't open the file, it's not compressed
+		return false
+	}
+
+	// check if the file is an archive
+	format, _, err := archiver.Identify(source, file)
+	if err != nil {
+		return false
+	}
+
+	// if the format is nil, it's not an archive
+	if format == nil {
+		return false
+	}
+
+	// if the format is an extractor, it's an archive
+	_, ok := format.(archiver.Extractor)
+	return ok
+}
+
+// IsJar checks if the file is a jar
+func IsJar(source string) bool {
+	return strings.ToLower(path.Ext(source)) == ".jar"
+}
```

Comment:
When/why did this come up? Do we support `.jar`s as "archives"?
And if we do, checking the extension might not always be good enough, better to check file type by magic number and add a test case with a proper jar file.